Documentation for OQL expressions

pijuskri opened the pull request:
- Documentation for OQL expressions following the new structure
- Includes aggregates and parameters
- Needs team review
- To be merged with other OQL documentation MR's

pijuskri committed:
Combine existing content into oql-expressions doc

pijuskri committed:
Parameter example improvements

pijuskri committed:
Aggregate documentation

pijuskri committed:
Aggregate documentation, examples

dahfjkg commented on content/en/docs/refguide/modeling/data/oql/oql-expressions.md:
````diff
@@ -0,0 +1,221 @@
+
+```sql
+  COUNT ( * )
+  | { COUNT | AVG | MAX | MIN | SUM } ( [ DISTINCT ] attribute_path )
````
Let's add `STRING_AGG` here

dahfjkg commented on content/en/docs/refguide/modeling/data/oql/oql-expressions.md:
````diff
@@ -0,0 +1,221 @@
+
+The sum of all products in stock:
+```sql
+SELECT SUM(Stock) AS StockSum FROM Sales.Product
````
This example is repeated twice

dahfjkg commented on content/en/docs/refguide/modeling/data/oql/oql-expressions.md:
```diff
@@ -0,0 +1,221 @@
+|:--------:|
+|    54    |
+
+Using `MAX` on a string column gives the last alphabetically sorted name:
```
TBH, this is probably too technical

pijuskri replied:
Agreed, i dont think the example is neccesary for more advanced users and this behaviour is mentioned in the documentation

dahfjkg commented on content/en/docs/refguide/modeling/data/oql/oql-expressions.md:
````diff
@@ -0,0 +1,221 @@
+    Active = true
+```
+
+If the value of `$param` is not provided, the query will be equivalent to:
````
I think this is not entirely correct. If a parameter is empty, that part of the expression should be replaced with true, like this:
```
WHERE
    true 
    AND
    Active = true
```
This is even more relevant in case of `OR`, so maybe we should use `OR` in our examples

pijuskri committed:
Correct parameter examples, remove bad examples

MarkvanMents committed:
Move oql-expressions to domain-model/oql

MarkvanMents committed:
Proofread oql-expressions

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+
+An OQL expression is a query building block that returns a value or a list of values. Expressions can be one of the following:
+
+* a constant
```
Can this also be a list of constants? Or can only some expressions return lists?

pijuskri replied:
There can be a list of of constants, but only when used with the `IN` operator. Don't think that single use case is important enough to mention here. Some expressions (like subquery) can return multiple values, so its useful to mention that here.

MarkvanMents replied:
OK, Thanks.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+* a combination of attribute names, constants, and functions connected by operators
+* a subquery
+
+OQL expressions can be used in `WHERE`, `SELECT`, `GROUP BY`, `UNION`, and `JOIN` clauses.
```
Should probably link to the OQL Clauses doc here.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+
+* as an attribute in a `SELECT` clause
+* as a condition in a `HAVING` clause
+* When combined with a `GROUP BY` clause, aggregations can be used with any column normally available in `SELECT`, not just those specified in the `GROUP BY` clause. Aggregation is performed over groups separately in this case.
```
I don't understand this at all - and there is no example. How can you put an aggregation into a GROUP BY clause if the column isn't mentioned in the GROUP BY clause?

pijuskri replied:
I misformated this. It should not be part of the "ways to use aggregations" list. 
The comment about group by is for aggregates the `SELECT` clause. Ill clarify this

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+|:------------:|
+|      4       |
+
+The same result can be retrieved by using `COUNT` on a single table:
```
On a single **attribute**?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+| Tomatoes | 44    |
+| Tomatoes | NULL  |
+
+#### COUNT
```
Should these examples be in the same order as we explain the expressions (i.e. with AVG first)?

pijuskri replied:
Yes, did not notice the order, will change it.

MarkvanMents replied:
I've reordered them as the initial description has it ordered COUNT (*) and then COUNT | AVG | MAX - so it seems better to present the descriptions in this order.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+
+| ProductCount |
+|:------------:|
+|    34.333    |
```
This is only right because one of the **Tomatoes** rows has NULL. If it had a number in then the average would be wrong 😀 because we aren't filtering on distinct product names.

pijuskri replied:
Thats true. Making this query distinct would require a subquery, so a change of the description is better.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+|:-----------------------------:|
+| Cheese,Milk,Tomatoes,Tomatoes |
+
+#### Distinct
```
I think we should link to the clauses doc here.
Also, DISTINCT isn't an expression - the example is about COUNT so I think we should move this to the COUNT examples.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+
+Parameters are external variables that are referenced to by name in an OQL query. To use a defined parameter in a query, prepend the `$` sign to the name of the parameter. 
+
+If you use undefined in `IN` and `LIKE` comparison expressions, the condition always returns `true`. In other cases, undefined parameters cause an exception.
```
We don't have any documentation for `IN` and `LIKE`. Are these expressions going to be added here, or will they be in a different document?

pijuskri replied:
Yes they will be documented in `oql-expression-syntax`

MarkvanMents replied:
OK, thanks.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
````diff
@@ -0,0 +1,264 @@
+    Job = 'Sales'
+```
+
+If the value of `$param` is not provided, the query will be equivalent to:
````
Not sure exactly what "not provided" means in this context.
Does it mean that the parameter has not been created and therefore is not in the current scope, or can you create a variable but still not initialize it.
If it is the first, then perhaps say "If the parameter `$param` has not been created"?

pijuskri replied:
The query can reference a parameter and then the parameter mapping needs to be provided when calling the query, say in a java action. I wouldn't say "created" is the best word to use here, parameters are quite temporary.

MarkvanMents replied:
So could we say
```suggestion
If the value of `$param` is not provided as a parameter to the query, the query will be equivalent to:
```

pijuskri replied:
Yes that seems more clear

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+
+## Parameters
+
+Parameters are external variables that are referenced to by name in an OQL query. To use a defined parameter in a query, prepend the `$` sign to the name of the parameter. 
```
Do we have some documentation on how these are created - I guess they are just Java variables, but how does that work in data sets?
Perhaps this would become obvious if I was actually creating some OQL?

pijuskri replied:
The best we have now is the datasets documentation. There arent any visuals or examples there and honestly I dont know how that works either.
For java actions its partially explained in the [API docs](https://apidocs.rnd.mendix.com/8/runtime/com/mendix/systemwideinterfaces/connectionbus/requests/types/ITextGetRequest.html#createParameterMap()).

MarkvanMents replied:
Thanks for the explanation. I've added a link to the API Docs.

MarkvanMents reviewed (commented):
Hi @pijuskri and @dahfjkg 
I've proofread the whole of this document.
I think the style matches the clauses document quite well, I'm not worried that they will seem too different.
I've asked some questions for clarification. 
I'll go back to the clauses document while you digest my changes and come back to me.
Let me know if you want to set up a meeting.

MarkvanMents reviewed (commented):
I've responded to some of your replies.

pijuskri committed:
Improve wording, change example order

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,264 @@
+
+OQL expressions can be used in `WHERE`, `SELECT`, `GROUP BY`, `UNION`, and `JOIN` clauses. For more information, see [OQL clauses](/refguide/oql-clauses/).
+
+## Aggregations
```
I'm referring to this in Clauses using tag #aggregates. Please add it here

pijuskri committed:
Add reference tag to aggregates section

pijuskri committed:
Align definition of expressions with oql clauses doc

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,265 @@
+* a constant
+* a function
+* a system variable
+* a combination of attribute names, constants, system variables and functions connected by operators
```
A subquery can also be a part of "a combination of attribute names, constants, ...", so I'd put this line last and mention subquery in it

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,265 @@
+* a combination of attribute names, constants, system variables and functions connected by operators
+* a subquery
+
+OQL expressions can be used in `WHERE`, `SELECT`, `GROUP BY`, `UNION`, and `JOIN` clauses. For more information, see [OQL clauses](/refguide/oql-clauses/).
```
1. I think `HAVING` too;
2. For `JOIN` clauses, I'd rephrase it as "and `ON` conditions of `JOIN` clauses"

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
```diff
@@ -0,0 +1,265 @@
+|:------------:|
+|      3       |
+
+There are duplicate values in the `Name` column, which might not want to be counted separately. [Distinct](/refguide/oql-clauses/#Distinct) can be used to get the number of unique rows:
```
Let's explain how `DISTINCT` works with `AVG` and `SUM` too

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
````diff
@@ -0,0 +1,265 @@
+
+```sql
+  COUNT ( * )
+  | { COUNT | AVG | MAX | MIN | SUM | STRING_AGG } ( [ DISTINCT ] attribute_path )
````
This syntax is not entirely correct: `DISTINCT` is allowed with all functions except `STRING_AGG`

pijuskri committed:
Handle review comments

Distinct examples, Expression definition, use cases, aggregate syntax, example aliases

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expressions.md:
````diff
@@ -100,7 +103,17 @@ SELECT AVG(Stock) AS ProductCount FROM Sales.Product
+There are duplicate values in the `Stock` column, which can be ignored by using [DISTINCT](/refguide/oql-clauses/#Distinct). The query below returns the average unique stock:
+
+```sql
+SELECT AVG(Distinct Stock) AS ProductCount FROM Sales.Product
````
Let's use more representative aliases for different aggregate functions. Both in queries and in result tables.

dahfjkg commented:
Passed technical review

MarkvanMents committed:
Proofread and confirm review changes

dahfjkg commented:
This can be closed as it is merged into https://github.com/mendix/docs/pull/8431

MarkvanMents commented:
No unresolved comments. Merged into #8431 so closing this PR.
